chemist/src/main/scala/Sharder.scala

```diff
@@ -24,15 +24,48 @@ import scalaz.std.set._
 trait Sharder {
   /**
    * provide the new distribution based on the result of calculating
-   * how the new set should actually be distributed. main benifit here
+   * how the new set should actually be distributed. main benefit here
    * is simply making the operations opaque (handling missing key cases)
    *
-   * Returns values of this function represent two things:
-   * 1. `Seq[(Flask,Target)]` is a sequence of targets zipped with the flask it was assigned too.
-   * 2. `Distribution` is that same sequence folded into a `Distribution` instance which can
-   *     then be added to the existing state of the world.
+   * Returned distribution includes all previous assignments as well as assignments for nee targets.
    */
-  def distribution(s: Set[Target])(d: Distribution): (Seq[(Flask,Target)], Distribution)
+  def distribution(s: Set[Target])(d: Distribution): Distribution
+}
+
+trait PartialSharder {
```

Comment:
Sample use case is defined below. 

I think it can be generalized further but i did not make my mind on what will be best approach. For now i only had one use case (ensure flasks own streams are stay on the flask).

Other possible extensions - app or placement aware allocation strategies. 
E.g. do something special for kafka to ensure "heavy" streams are not assigned to the same flask or do availability zone or VPC aware placement.